### DISCUSSION TITLE
CLOG 1.6 - Dev Journal

### DISCUSSION THREAD

**Discussion by rabbibotton** [rabbibotton/clog#183] [General]
The 1.6 target is Projects and better property editing

**Comment by rabbibotton** [comment]
I have made a change to long polling.  Using :long-poll-first :extend gives the previous behavior, long polling continues to run until the on-new-window handler is completed. :long-poll-first t will the stop long polling on first query made to browser. Leaving long-polling on creates the appearance that the site is still loading and its chief goal is for search engine crawling that is better served by keeping it short and the main data upfront.

**Reply by rabbibotton** [reply]
I changed the default behavior back for the :long-poll-first parameter but now can specify how many requests made to browser before stopping to long-poll.

**Comment by rabbibotton** [comment]
(defgeneric replace-element (clog-element new-clog-element)
  (:documentation "CLOG-ELEMENT and its children will we removed from DOM
but not destroyed and NEW-CLOG-ELEMENT will replace it in same location.
DETROY CLOG-ELEMENT if need to clear from browser memory.
Returns CLOG-ELEMENT"))

**Comment by rabbibotton** [comment]
(defgeneric swap-element-by-id (html-id clog-element)
  (:documentation "Replace an existing element on page with HTML-ID with
CLOG-ELEMENT and set CLOG-ELEMENT to html-id. Returns CLOG-ELEMENT"))

**Comment by rabbibotton** [comment]
Added a simple thread viewer under tools in builder

**Comment by rabbibotton** [comment]
basic system browser

<img width="816" alt="image" src="https://user-images.githubusercontent.com/11616789/181387893-1923fffc-61a7-4719-a3fc-d4824d470cf6.png">

**Reply by rabbibotton** [reply]
Added class only mode and more descriptive entries

<img width="810" alt="image" src="https://user-images.githubusercontent.com/11616789/181406237-4bd355c8-5ced-44af-9caa-042fd2a30781.png">

**Reply by rabbibotton** [reply]
Now a search box, so now really usable. I think next to add ability to change and save.

<img width="748" alt="image" src="https://user-images.githubusercontent.com/11616789/181411993-8f043c36-d66a-4ddd-80c6-adf62e8b8eac.png">

**Comment by aykaramba** [comment]
Wow.

**Reply by rabbibotton** [reply]
I hope to expand on it more in the future. For now is very basic.

**Comment by rabbibotton** [comment]
For now final version - very functional I love it :)
Can change files now in the browser, save them. Eval the file or eval a selection.

![image](https://user-images.githubusercontent.com/11616789/181613141-41ad9938-84ed-4ac1-a07d-dc0b61c81122.png)

**Reply by rabbibotton** [reply]
Now choosing an item brings you directly to the correct def* depending on type. Eventually I'll have the M-. functionality go directly to a system browser.

**Reply by rabbibotton** [reply]
Once you start builder, you can also launch a system browser as it's own window. This is very convenient especially if looking to code with it.

**Comment by rabbibotton** [comment]
May not have time to complete all but I have a new ASDF System Browser - I renamed the System Browser to Package Browser.
Once done you will be able to double click the files to load them. I think we can add a reload - to reload the system from sources.
I want to extend asdf to handle clog files as well. I think also want to use asdf + editions as the main project format overall.
I also moved both browser to the project menu.

<img width="650" alt="image" src="https://user-images.githubusercontent.com/11616789/181852655-47da4f16-8434-4731-be77-04301b1b7e85.png">

**Reply by rabbibotton** [reply]
Ok, so now that is all working - double click files to open and sub directories for those systems that have and can double click dependencies.

<img width="823" alt="image" src="https://user-images.githubusercontent.com/11616789/182011731-d9fc99c6-418c-4789-bfe6-d35fc0bdbeb6.png">

**Comment by rabbibotton** [comment]
New feature - clicking on the side panels unlocks/locks them so they auto slide in and out with mouse. Much more functional for using the builder as a general IDE

**Comment by rabbibotton** [comment]
Much more functional window management in clog-gui you can now click any place on overlapping window to bring to front, and the select window list does not maximize all windows.

**Comment by rabbibotton** [comment]
The builder now extends ASDF to include :clog-file for any defsystem that includes :defsystem-depends-on (:clog). You still currently have to specify the :file component also for the generated lisp file. All of the templates have been updated to include the line.

You already can use the ASDF Browser like a project tool and can double click on clog files to open in builder as panels for editting and of course .lisp files and .asd files etc.

This is a typical asd layout from the templates.(notice - we use the base system for deployment, and the /tools for design and that is where the clog files would be and any builder plug-ins, the base system would have the lisp files and any the plugins without the /tools builder portion):

```
(asdf:defsystem #:syspractice
  :description "New CLOG System"
  :author "some@one.com"
  :license  "BSD"
  :version "0.0.0"
  :serial t
  :depends-on (#:clog #:clog-ace) ; add clog plugins here as #:plugin for run time
  :components ((:file "syspractice")
               (:file "sys1")))

(asdf:defsystem #:syspractice/tools
  :defsystem-depends-on (:clog)
  :Depends-on (#:syspractice #:clog/tools #:clog-ace/tools) ; add clog plugins here as #:plugin/tools for design time
  :components ((:clog-file "sys1")))
```

<img width="811" alt="image" src="https://user-images.githubusercontent.com/11616789/182050749-3eb71c9b-01b2-45c1-a63e-70f27760fe10.png">

**Comment by rabbibotton** [comment]
Renamed the Package Browser back to System Browser because it handles more now.

M-. now uses the System Browser instead of requiring emacs.

When choosing for packages "All" the browser does an apropos style find of the entire image and the search box requires hitting enter. Once you drill down to a package the search is instant still.

**Comment by rabbibotton** [comment]
The first pic of the project window and a project asd file. I don't know if will finish today or tomorrow but working on adding modifying asd files (add .lisp, add .clog, delete)

<img width="765" alt="image" src="https://user-images.githubusercontent.com/11616789/182277005-a45a6794-c228-40f9-be62-26bd33767bf5.png">

**Comment by rabbibotton** [comment]
Three things if any one wants to work on (these are not trivial) as I won't get to it for some time:

1. At some point need to extend asdf so that the :clog-file asdf components give a warning if are more up to date then their generated .lisp component. I'll provide in future a batch compile .clog -> .lisp. Would be cool if that extension to asdf could call the tool once ready by file. (This is very doable just time consuming to go through asdf code and decode the docs for me for now)
2. We need a way to switch the static-route dir (www/) being used by the webserver on the fly.
3. A way to build a static-route dir (www/) in lisp code (that also doesn't play with the file system like with sym links etc) this way we can build up the various resources used by plugins etc.

**Comment by rabbibotton** [comment]
Projects are ready. You can only add/remove lisp or clog files from properly setup .asd files for CLOG. Although you can use other .asd files it will disable use of add/remove. All templates produce correct asd projects, but it is easy enough to do by hand or retrofit - you must have at least the base defsystem and /tools, /tools also must include :defsystem-depends-on (:clog). Since clog is altering the .asd's tooled for clog, they are going to loose formatting and any comments.

<img width="710" alt="image" src="https://user-images.githubusercontent.com/11616789/182508210-64b327d1-ccbf-48f8-bd43-cd3dcde3571b.png">

**Reply by rabbibotton** [reply]
oh if you want (clog-tools:clog-builder :project :your-proj) makes :your-proj the default project

**Comment by rabbibotton** [comment]
autofill now based on package of source file or panel
eval now finds the package of source file and evals in that package

**Comment by rabbibotton** [comment]
The project window now handles dependencies and plugins so that is a project has a plugin it is installed also in the control palette

<img width="708" alt="image" src="https://user-images.githubusercontent.com/11616789/182958633-68247f71-b22b-4950-b74d-392936c82bbd.png">

**Comment by rabbibotton** [comment]
alt/cmd-[ to evaluate current form in any ace editor. If the current editor has an in-package (of course should) it will be evaluated in that package otherwise clog-user

**Comment by rabbibotton** [comment]
Now there are "eval form" buttons wherever there were other eval buttons.

**Comment by rabbibotton** [comment]
Now uses the asdf entry-point to store the entry point so can do a "run" on your project from gui. Templates updated to start with correct entry-point, you can add yours or change

<img width="641" alt="image" src="https://user-images.githubusercontent.com/11616789/183219060-551a6798-b65d-4ddb-a55e-dc1e605c2829.png">

**Comment by rabbibotton** [comment]
The status bar that shows the form parameters now works better to match package
Eval results (except when done with cmd/alt-[) are now returned in an editor window.

**Comment by rabbibotton** [comment]
Added "Callers and Callee" tools. They really could easily be made in to cool looking tools not just a list, but good enough for the moment.

**Comment by rabbibotton** [comment]
Added create-clog-webgl that will get a WebGL 2 context from a canvas control. Looks like I am going to get started on the WebGL bindings this week and next.

**Reply by rabbibotton** [reply]
First CLOG WebGL Image :) - tutorial 34
lots more to go

<img width="1139" alt="Screen Shot 2022-08-09 at 9 11 19 PM" src="https://user-images.githubusercontent.com/11616789/183790119-00ab3c41-4db1-4add-88f4-9a973cf6f5d7.png">

**Reply by rabbibotton** [reply]
Tutorial 35 - 3D object
Don't plan on pursuing more on this. Mainly did as proof of concept. The 2D Canvas also need some love and in some future iteration will flesh out more of both.

![image](https://user-images.githubusercontent.com/11616789/184784790-deb3c9ce-f93d-4d58-a017-d12a6a2ba649.png)

**Comment by rabbibotton** [comment]
clog ace and terminal can now use local version when present and local copies are now added to repo for speed and to remove need for internet. I will also be enhancing the lisp mode on ace soon.

**Reply by rabbibotton** [reply]
improved it a bit, multi-line comments, add a few more keywords. I will need to redo it completely at some point.

**Reply by rabbibotton** [reply]
made a small change in Ace's ext-language_tool.js and we now have full autocomplete for any package using swank,

var ID_REGEX = /[a-zA-Z_0-9\:\$\-\u00A2-\u2000\u2070-\uFFFF]/;

Added \:

**Comment by rabbibotton** [comment]
Probably the most awesome thing in the upcoming 1.6 - Builder Panel Project Templates

This first one is deceptively simple - instantly build multi page websites / GUIs - your first panel is centered on web browser, the next button - replaces the current panel with the next one using (replace-element panel (create-next-panel panel)), and the next panel does the same to create a new main page to replace itself.

This needs a video, but this allows you to dev completely in VB / Delphi style.

**Reply by rabbibotton** [reply]
Here is a vid - not so great audio. Tomorrow will make something of better quality

https://www.reddit.com/user/dbotton/comments/wqerh3/clog_builder_panel_templates/

**Comment by rabbibotton** [comment]
Full screen/browser Panel Project

<img width="1510" alt="image" src="https://user-images.githubusercontent.com/11616789/185469079-3140176c-4bd7-4df8-a45c-40784fa719f4.png">

**Comment by rabbibotton** [comment]
I think going to move the property work to version 1.7

**Comment by janegwaww** [comment]
very expect a web emacs.
